Src/Base/AMReX_GpuFuse.H

```diff
@@ -0,0 +1,280 @@
+    }
+}
+
+template <typename Lambda>
+void
+PutLambdaDtor (LambdaDtor& lambda_dtor)
+{
+    lambda_dtor = &DestroyLambda<Lambda>;
+}
+
+class Fuser
+{
+public:
+
+    Fuser ();
+    ~Fuser ();
+    Fuser (Fuser const&) = delete;
+    Fuser& operator= (Fuser const&) = delete;
+
+    template <typename F>
+    void
+    Register (Box const& bx, F&& f)
+    {
+        AMREX_ASSERT(!OpenMP::in_parallel());
+        if (bx.isEmpty()) return;
+        using Lambda = typename std::decay<F>::type;
+        Register_doit<Lambda>(bx, 0, f);
+        PutLambda3DLauncher<Lambda>(m_helper_buf+(m_nlambdas-1));
+    }
+
+    template <typename F>
+    void
+    Register (Box const& bx, int ncomp, F&& f)
+    {
+        AMREX_ASSERT(!OpenMP::in_parallel());
+        if (bx.isEmpty()) return;
+        using Lambda = typename std::decay<F>::type;
+        Register_doit<Lambda>(bx, ncomp, f);
+        PutLambda4DLauncher<Lambda>(m_helper_buf+(m_nlambdas-1));
+    }
+
+    template <typename T, typename F, typename M=amrex::EnableIf_t<std::is_integral<T>::value> >
+    void
+    Register (T N, F&&f)
+    {
+        AMREX_ASSERT(!OpenMP::in_parallel());
+        AMREX_ASSERT(sizeof(T) <= sizeof(int) ||
+                     N < static_cast<T>(std::numeric_limits<int>::max()));
+        if (N <= 0) return;
+        using Lambda = typename std::decay<F>::type;
+        Register_doit<Lambda>(Box(), N, f);
+        PutLambda1DLauncher<Lambda>(m_helper_buf+(m_nlambdas-1));
+    }
+
+    void Launch ();
+
+    static Fuser& getInstance ();
+    static void Initialize ();
+    static void Finalize ();
+
+private:
+
+    static std::unique_ptr<Fuser> m_instance;
+
+    char* m_lambda_buf = nullptr;
+    std::size_t m_nbytes_used_lambda_buf = 0;
+    std::size_t m_nbytes_lambda_buf = 2*1024*1024;
+
+    FuseHelper* m_helper_buf = nullptr;
+    std::size_t m_nhelpers_buf = 1024;
+
+    Vector<LambdaDtor> m_dtor_buf;
+
+    int m_nlambdas = 0;
+
+    void resize_lambda_buf ();
+    void resize_helper_buf ();
+
+    template <typename Lambda>
+    void
+    Register_doit (Box const& bx, int ncomp, Lambda const& f)
+    {
+        if (m_nlambdas >= m_nhelpers_buf) {
+            resize_helper_buf();
+        }
+
+        int ikernel = m_nlambdas++;
+        new (m_helper_buf+ikernel) FuseHelper;
+        m_helper_buf[ikernel].m_offset = m_nbytes_used_lambda_buf;
+        m_helper_buf[ikernel].m_bx = bx;
+        m_helper_buf[ikernel].m_N = ncomp;
+
+        static_assert(std::alignment_of<Lambda>::value <= Arena::align_size,
+                      "GpuFuse: alignment error");
+        if (m_nbytes_used_lambda_buf + Arena::align(sizeof(Lambda)) > m_nbytes_lambda_buf) {
+            resize_lambda_buf();
+        }
+        char* p_lambda = m_lambda_buf + m_nbytes_used_lambda_buf;
+        std::size_t sz_lambda = PutLambda<Lambda>(f, p_lambda);
+        m_nbytes_used_lambda_buf += sz_lambda;
+        AMREX_ASSERT(m_nbytes_used_lambda_buf <= m_nbytes_lambda_buf);
```

Comment:
If I'm reading this correctly, this assertion is unnecessary, given the if statement check above and the resize_lambda_buf function does what it's designed to do. If it's possible the resize has the possibility of not increasing the buffer enough, would it make sense to pass the resize a minimum size to add to guarantee success?

Reply:
Yes, you could argue the assertion is unnecessary.  Like a lot of assertions we do, it's just there to rule out possibilities if the program does not actually do what we think it is supposed to do.   And the assertion is there for debug build only.  Yes, it's pretty much impossible the resize does not increase the buffer enough.